### ISSUE TITLE
tf.io.gfile / GCS fails to work on OpenSUSE

### ISSUE THREAD

**Issue by kazimuth** [tensorflow/tensorflow#40065]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux OpenSUSE Tumbleweed
- TensorFlow installed from (source or binary): Binary (conda)
- TensorFlow version (use command below): unknown 2.1.0
- Python version: 3.7.5
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**
```python
import tensorflow as tf
tf.io.gfile.listdir("gs://some-bucket") # replace w/ bucket of your choice
```

This code gives an error:
```
2020-06-01 15:43:56.684531: W tensorflow/core/platform/cloud/google_auth_provider.cc:178] All attempts to get a Google authentication bearer token failed, returning an empty token. Retrieving token from files failed with "Unavailable: Error executing an HTTP request: libcurl code 77 meaning 'Problem with the SSL CA cert (path? access rights?)', error details: error setting certificate verify locations:
  CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: none". Retrieving token from GCE failed with "Aborted: All 10 retry attempts failed. The last failure: Unavailable: Error executing an HTTP request: libcurl code 6 meaning 'Couldn't resolve host name', error details: Couldn't resolve host 'metadata'".
```

After that it hangs for a while, and then raises a `NotFoundError`.

I believe this is because the libcurl packaged with tensorflow doesn't know where to find the ca-certificates bundle file on OpenSUSE, which is at `/etc/ssl/ca-bundle.pem` rather than `/etc/ssl/certs/ca-certificates.crt`. Also, I installed through miniconda so there's another equivalent file at `$CONDA_PREFIX/ssl/cacert.pem`. Neither of these seems to be found by tensorflow.

[This code](https://github.com/tensorflow/tensorflow/blob/5597c17b6a677be5264ebda7cc31404f0ae8a434/tensorflow/core/platform/cloud/curl_http_request.cc#L129-L132) suggests that the bundle file's location can be customized with the `CURL_CA_BUNDLE` env variable. However, this doesn't change the behavior as far as i can tell; the error is still raised.

**Describe the expected behavior**
It should list the contents of the bucket.

**Comment by kazimuth** [comment]
Ping on this

**Comment by mihaimaruseac** [comment]
Can you run the following please?

```
strace python-c "import tensorflow as tf; tf.io.gfile.listdir("gs://some-bucket") &> log1.txt
strace CURL_CA_BUNDLE=path/to/good/certs python-c "import tensorflow as tf; tf.io.gfile.listdir("gs://some-bucket") &> log2.txt
```

And then attach the `log*.txt` files in a reply? You can use multiple paths for the certs and attach multiple logs.

Note that since we only build and test on ubuntu, it might be slow to solve the issue. We don't have enough headcount to solve issues on all operating systems, unfortunately.

**Comment by qiyubio** [comment]
We had a similar problem with

CentOS 7
Python 3.7.5
tensorflow 2.1.0

Manually set CURL_CA_BUNDLE do not seems work.

**Comment by Flamefire** [comment]
I ran into this again on RHEL 7 with TF 2.2 and 2.3 and dug a bit:

- CURL_CA_BUNDLE is for the binary only, not the library. See https://github.com/curl/curl/blob/0d0537aeae2acdbdb43c6aab834abf84ca7ba247/docs/SSLCERTS.md
- The CMake files for CURL correctly search for an existing cert: https://github.com/curl/curl/blob/13030d08ad392e3a6c139400f0dfa827f1e9f2a8/CMakeLists.txt#L830-L835
- TensorFlow hard-codes the location at https://github.com/tensorflow/tensorflow/blob/50db873b7c593ecb573cb44a8cf0c91292cfc515/third_party/curl.BUILD#L487

Hence for TF to work the CA cert MUST be at `/etc/ssl/certs/ca-certificates.crt` and it is NOT possible to change that.

The only solutions are to use the system curl (see TF_SYSTEM_LIBS, but that is partly broken), make TF use the CMake install of curl (but I doubt that is gonna happen) or implement the location detection in Bazel

**Comment by Saduf2019** [comment]
@kazimuth 
Can you please verify if this is still an issue on later tf versions.

**Comment by Flamefire** [comment]
@Saduf2019 Yes the problem is still present because the problematic code hasn't changed: https://github.com/tensorflow/tensorflow/blob/7e193718033d18a159fcae75be4dd0a59c32b897/third_party/curl.BUILD#L572

**Comment by Flamefire** [comment]
This is still an issue, please do not close

**Comment by wanderlustzoe** [comment]
ping @jhseu ... this is still an issue. Anyone?

**Comment by swghosh** [comment]
Gentle ping. This problem is still around atleast on Fedora 35 with latest version of TensorFlow. (expecting the same to occur on other RPM based distros)

I could use this workaround for the time being:
```sh
sudo ln -s /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-certificates.crt
```

**Comment by sheetalgiri** [comment]
fixing CURL certificates path as suggested in the [code](https://github.com/tensorflow/tensorflow/blob/5597c17b6a677be5264ebda7cc31404f0ae8a434/tensorflow/core/platform/cloud/curl_http_request.cc#L129-L132) mentioned above seems to work for me for Tensorflow 2.4.1 and Fedora 28
added the following in .bashrc 
> export CURL_CA_BUNDLE=/etc/ssl/certs/ca-bundle.crt

**Comment by mihaimaruseac** [comment]
Can you send a PR please?

**Comment by mohantym** [comment]
Hi @kazimuth !

We are checking to see whether you still need help in this issue . 
Have you checked in the newer versions. I am not able to replicate this issue in [2.9,](https://colab.sandbox.google.com/gist/mohantym/74953b708ba0c6a62bd0d56bdc2c8119/git_40065.ipynb#scrollTo=KSQ1zihSwBnK) [2.10](https://colab.sandbox.google.com/gist/mohantym/5f5416629d1afed409489f00ede0963b/git_40065.ipynb#scrollTo=KSQ1zihSwBnK) and [2.11](https://colab.sandbox.google.com/gist/mohantym/b9621dce39c0b05b19623520f8e58e85/git_40065.ipynb#scrollTo=KSQ1zihSwBnK).

Could you let us know from your side.

Thank you!

**Comment by mihaimaruseac** [comment]
Colab is not using OpenSUSE. There has been no PR here, so this is still valid.

**Comment by mathemakitten** [comment]
Just confirming that I also ran into this with gfile in TF 2.11.0 on an AWS cluster running Amazon Linux 2 (RHEL) and `sudo ln -s /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-certificates.crt` did fix it.

**Comment by butybone** [comment]
running:
sudo ln -s /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-certificates.crt
but got an error:
ln: failed to create symbolic link '/etc/ssl/certs/ca-certificates.crt': File exists

**Comment by NicoleYY77** [comment]
I'm running into the same issue... Is there a final solution to it?

**Comment by mihaimaruseac** [comment]
This still needs a PR based on https://github.com/tensorflow/tensorflow/issues/40065#issuecomment-1095048218

Alternatively, you can try the workaround from https://github.com/tensorflow/tensorflow/issues/40065#issuecomment-1053667253

**Comment by Venkat6871** [comment]
Hi @kazimuth ,
Thank you so much for filing this bug request! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.To avoid duplicate work, please leave a comment here if you decide to take it on!
Thank you again for your interest and we look forward to seeing your contributions!
